tour/programs-functions.md

```diff
@@ -0,0 +1,207 @@
+
+### Pulumi execution order.
+
+`pulumi` uses `node` to executes a pulumi-application.  During execution, when a call to `createLambdaFunction` is encountered, the function is converted to a Lambda at that point in execution.  That means that if the function captures any state, then the value that is captured will be whatever it was at the point in time.  For example consider the following two programs:
+
+```ts
+let obj = { a: 1, b: 2 };
+
+const lambda: aws.lambda.Function = serverless.function.createLambdaFunction("mylambda", 
+    async () => {
+        console.log(obj);
+    });
+
+obj = { a: 3, b: 4 };
+```
+
+
+```ts
+let obj = { a: 1, b: 2 };
+obj = { a: 3, b: 4 };
+
+const lambda: aws.lambda.Function = serverless.function.createLambdaFunction("mylambda", 
+    async () => {
+        console.log(obj);
+    });
+```
+
+When `pulumi` starts executing `createLambdaFunction` it will analyze the JavaScript function code and will see that it uses the 'obj' value.  At that point in time it will use whatever the value is currently to serialize over.  So, in the first example, it will serialize the value `{ a: 1, b: 2 }`, even though right after executing createLambdaFunction the program code will update that value to `{ a: 3, b: 4}`.  In the second example, the code will see the `{ a: 3, b: 4 }` value and will serialize that into the cloud-runtime code.  
+
+Importantly: there are subtleties here with Promise-like values.  When `pulumi` encounters a Promise value that it needs to serialize into the code for a Lambda, it will actually 'await' that Promise.  During that 'await', 'node' can then execute more of the program application code.  This means that later code may execute which then changes a value which is captured by the JavaScript function.  If `pulumi` then serialized that value after serializing the Promise, then it may see the mutated value.
+
+So, in the following code:
+
+```ts
+let obj = { a: 1, b: 2 };
+let pr = new Promise((resolve, reject) => { /*...*/ });
+
+const lambda: aws.lambda.Function = serverless.function.createLambdaFunction("mylambda", 
+    async () => {
+        console.log(pr);
+        console.log(obj);
+    });
+
+obj = { a: 3, b: 4 };
+```
+
+it may be the case that the value `{ a: 1, b: 2}` or `{ a: 3, b: 4}` is serialized depending on the order that things are serialized in.
+
+For this reason, it is highly recommended that code not capture values which are also mutated in code.  It is much safer and easier to reason about immutable values that are captured in code.
```

Comment:
I would move the sentence at line 207 to line 162 ... iow, the first sentence on execution order ... it's the headline, and while the explanation is great -- don't bury the lead -- "It is highly recommended that Pulumi programs do not capture values which are also mutated in code."

Reply:
👍